p4-16/psa/PSA.mdk

```diff
@@ -756,45 +778,43 @@ the contents of several metadata fields in the struct
                 ostd.clone_class_of_service, after which it will start
-                egress processing with egress_port=ostd.clone_port (which
-                might be PORT_CPU to send the clone to the control CPU).
-            The contents of the packet will be the same as the packet
-                as it comes out of the egress deparser, prepended with any
-                headers created by clone.emit() calls in the egress
-                deparser apply block.
+                egress processing.
         } else {
             // Do not create a clone.  Recommended to log error about
             // unsupported ostd.clone_port value.
         }
     }
-    // Continue below, regardless of whether a clone was created or not.
+    // Continue below, regardless of whether a clone was created.
+    // Any clone created above is unaffected by the code below.
     if (ostd.drop) {
         drop the packet
         return;   // Do not continue below.
     }
     if (ostd.recirculate) {
         recirculate the packet, i.e. it will go back to starting with the
             ingress parser;
-        // TBD: Specify if anything is different about the recirculated
-        // packet vs. other copies that might be made below.
         return;   // Do not continue below.
     }
 
     // The value istd.egress_port below is the same one that the
-    // packet began its Egress processing with, as decided during
-    // Ingress processing for this packet.  The Egress control block
-    // is not allowed to change it.
+    // packet began its egress processing with, as decided during
+    // ingress processing for this packet.  The egress code is not
+    // allowed to change it.
     enqueue one packet for output port istd.egress_port
 ```
 
-TBD: Need text defining, for each possible copy, exactly what the
-contents of the packet will be, and any differences between the values
-of the fields in the structs `psa_parser_input_metadata_t` and
-`psa_egress_input_metadata_t`, in the copy, as compared to the values
-seen for the packet that caused those copies to be made.
+TBD: Should it be possible to truncate a cloned packet differently
+than the normal packet that goes out?
 
-TBD: Should it be possible to truncate a cloned or recirculated packet
-differently than the normal packet that goes out?
+As for the handling of a packet after ingress processing, a PSA
+implementation may drop a packet after egress processing, even if the
+pseudocode above says that a packet will be sent.  For example, you
+may attempt to clone a packet after egress when the packet buffer is
+too full, or you may attempt to recirculate a packet when the ingress
+pipeline is busy handling other packets.  It is recommended that an
+implementation maintain counters of packets dropped, preferably with
+separate counters for as many different reasons as the implementation
+has for dropping packets outside the control of the P4 program.
```

Comment:
Similar comment as one I called out above -- there must be cases where a packet can be dropped even if the P4 program didn't request it.